ibis/expr/tests/test_analysis.py

```diff
@@ -160,13 +160,33 @@ def test_bad_join_predicate_raises():
     filter_pred = table['f'] > 0
     table3 = table[filter_pred]
 
-    with pytest.raises(com.ExpressionError):
-        table.inner_join(table2, [table3['g'] == table2['key']])
-
+    result = table.inner_join(table2, [table3['g'] == table2['key']])
+    repr(result)
```

Comment:
Currently doesn't raise, neither equals (at least with `assert_equal`)

Reply:
What are the semantics of this? It seems odd to allow a filtered table's column to be in the expression if the left side of the join is not an immediate child of it.

Reply:
I might missed the proper validator, but I'd like to make the following working:

```python
t1 = ibis.table({'key2': 'string', 
                 'key3': 'string',
                 'value1': 'double'})

t2 = ibis.table({'key2': 'string', 
                 'value2': 'double'})

t3 = ibis.table({'key3': 'string', 
                 'value3': 'double'})

joined = t1.inner_join(t2, ['key2'])[t1, t2.value2]
joined2 = joined.inner_join(t3, ['key3'])[t1, t2.value2, t3.value3]
```

ideally without intermediate projection

```python
joined = t1.inner_join(t2, ['key2']).inner_join(t3, ['key3'])[t1, t2.value2, t3.value3]
```